tools/mpremote/mpremote/transport_serial.py

```diff
@@ -95,29 +99,45 @@ def close(self):
         self.serial.close()
 
     def read_until(self, min_num_bytes, ending, timeout=10, data_consumer=None):
-        # if data_consumer is used then data is not accumulated and the ending must be 1 byte long
+        """
+        min_num_bytes: Return at least min_num_bytes.
+        ending: Return if 'ending' matches.
+        timeout [s]: Return if timeout between characters. None for infinite timeout.
+        data_consumer: Use callback for incoming characters.
+            If data_consumer is used then data is not accumulated and the ending must be 1 byte long
+
+        It is not visible for the caller why the function returned. It could be ending, timeout, min_num_bytes.
+        """
         assert data_consumer is None or len(ending) == 1
+        assert isinstance(timeout, (type(None), int, float))
 
-        data = self.serial.read(min_num_bytes)
-        if data_consumer:
-            data_consumer(data)
-        timeout_count = 0
+        data = b""
+        begin_char_s = time.monotonic()
         while True:
-            if data.endswith(ending):
+            if timeout and time.monotonic() > begin_char_s + timeout:
```

Comment:
What happens if `timeout` is 0?

Reply:
A `timeout` of 0 is still considered the same as `None`, ie no timeout.  That's not the original semantics, where 0 would mean "non blocking" (except for the initial read...).

Reply:
@hmaerki this comment still needs addressing.  The code must handle timeout=0 to mean non-blocking.

Reply:
Sorry to be pedantic here, but I think there is an edge case where the timeout is small but non-zero and the PC is slow (eg switches to a different task which for some time) and this means the timeout is reached here before the serial port is checked.

The serial port should always be checked at least once, before the timeout is checked.